src/system/PermissionsModule/Api/PermissionApi.php

```diff
@@ -0,0 +1,323 @@
+            // if $instance is empty, then there must be a component.
+            // Looking for best permission
+            foreach ($perms as $perm) {
+                // component check
+                if (!preg_match("=^$perm[component]$=", $component)) {
+                    continue; // component doestn't match.
+                }
+
+                // check that the instance matches :: or '' (nothing)
+                if (!(preg_match("=^$perm[instance]$=", '::') || preg_match("=^$perm[instance]$=", ''))) {
+                    continue; // instance does not match
+                }
+
+                // We have a match - set the level and quit
+                $level = $perm['level'];
+                break;
+            }
+
+            return $level;
+        }
+
+        // Normal permissions check
+        // there *is* a $instance at this point.
+        foreach ($perms as $perm) {
+
+            // if there is a component, check that it matches
+            if (($component != '') && (!preg_match("=^$perm[component]$=", $component))) {
+                // component exists, and does not match.
+                continue;
+            }
+
+            // Confirm that instance matches
+            if (!preg_match("=^$perm[instance]$=", $instance)) {
+                // instance does not match
+                continue;
+            }
+
+            // We have a match - set the level and quit looking
+            $level = $perm['level'];
+            break;
+        }
+
+        return $level;
+    }
+
+    /**
+     * Fix security string.
+     *
+     * @param string $string String.
+     *
+     * @return string
+     */
+    private function normalizeRegexString($string)
+    {
+        if (empty($string)) {
+            $string = '.*';
+        }
+        if (strpos($string, ':') === 0) {
+            $string = '.*' . $string;
+        }
+        $string = str_replace('::', ':.*:', $string);
+        if (strrpos($string, ':') === strlen($string) - 1) {
+            $string = $string . '.*';
+        }
+
+        return $string;
+    }
+
+    /**
+     * Translation functions
+     *
+     * Translate level -> name
+     *
+     * @param integer $level Access level.
+     *
+     * @return string Translated access level name.
+     */
+    public function accessLevelName($level)
+    {
+        $accessnames = [
+            ACCESS_NONE => $this->translator->__('No access'),
+            ACCESS_OVERVIEW => $this->translator->__('Overview access'),
+            ACCESS_READ => $this->translator->__('Read access'),
+            ACCESS_COMMENT => $this->translator->__('Comment access'),
+            ACCESS_MODERATE => $this->translator->__('Moderate access'),
+            ACCESS_EDIT => $this->translator->__('Edit access'),
+            ACCESS_ADD => $this->translator->__('Add access'),
+            ACCESS_DELETE => $this->translator->__('Delete access'),
+            ACCESS_ADMIN => $this->translator->__('Admin access'),
```

Comment:
this method is used once or twice in the core and once in News that I know of. (If that was your question)